TESTS/mbed_drivers/flashiap/main.cpp

```diff
@@ -178,15 +178,95 @@ void flashiap_program_error_test()
     TEST_ASSERT_EQUAL_INT32(0, ret);
 }
 
+void flashiap_timing_test()
+{
+    FlashIAP flash_device;
+    uint32_t ret = flash_device.init();
+    TEST_ASSERT_EQUAL_INT32(0, ret);
+    mbed::Timer timer;
+    int num_write_sizes;
+    unsigned int curr_time, byte_usec_ratio;
+    unsigned int avg_erase_time = 0;
+    unsigned int max_erase_time = 0, min_erase_time = (unsigned int) -1;
+    const unsigned int max_writes = 128;
+    const unsigned int max_write_sizes = 6;
+    const unsigned int max_byte_usec_ratio = 200;
+
+    uint32_t page_size = flash_device.get_page_size();
+    uint32_t write_size = page_size;
+
+    uint32_t end_address = flash_device.get_flash_start() + flash_device.get_flash_size();
+
+    printf("\nFlash timing:\n");
+    uint32_t sector_size = flash_device.get_sector_size(end_address - 1UL);
+    uint32_t base_address = end_address - sector_size;
+    timer.start();
+    for (num_write_sizes = 0; num_write_sizes < max_write_sizes; num_write_sizes++) {
+        if (write_size > sector_size) {
+            break;
+        }
+        uint8_t *buf = new (std::nothrow) uint8_t[write_size];
+        if (!buf) {
+            // Don't fail the test on lack of heap memory for the buffer
+            break;
+        }
+        memset(buf, 0x5A, write_size);
+        timer.reset();
+        ret = flash_device.erase(base_address, sector_size);
+        curr_time = timer.read_us();
+        avg_erase_time += curr_time;
+        TEST_ASSERT_EQUAL_INT32(0, ret);
+        max_erase_time = std::max(max_erase_time, curr_time);
+        min_erase_time = std::min(min_erase_time, curr_time);
+        uint32_t address = base_address;
+        unsigned int avg_write_time = 0;
+        unsigned int max_write_time = 0, min_write_time = (unsigned int) -1;
+        int num_writes;
+        for (num_writes = 0; num_writes < max_writes; num_writes++) {
+            if ((address + write_size) > end_address) {
+                break;
+            }
+            timer.reset();
+            ret = flash_device.program(buf, address, write_size);
+            curr_time = timer.read_us();
+            avg_write_time += curr_time;
+            TEST_ASSERT_EQUAL_INT32(0, ret);
+            max_write_time = std::max(max_write_time, curr_time);
+            min_write_time = std::min(min_write_time, curr_time);
+            address += write_size;
+        }
+        delete[] buf;
+        avg_write_time /= num_writes;
+        printf("Write size %6u bytes: avg %10u, min %10u, max %10u (usec)\n",
+                write_size, avg_write_time, min_write_time, max_write_time);
+        byte_usec_ratio = write_size / avg_write_time;
+        TEST_ASSERT(byte_usec_ratio < max_byte_usec_ratio);
+        write_size *= 4;
+    }
+
+    if (num_write_sizes) {
+        avg_erase_time /= num_write_sizes;
+        printf("\nErase size %6u bytes: avg %10u, min %10u, max %10u (usec)\n\n",
```

Comment:
still there are 3 prints that do not have any value until the test fails (if it fails, then printing erase size makes sense but only if it fails).

Can we remove all 3 printf from from this   test case (move them to assert msg rather)

Reply:
@0xc0170 I would like to keep these prints to monitor these times also on success tests. We suspect these times can change with wear of the flash. They will also help us to understand for every device the level of intrusiveness of NVStore and SOTP

Reply:
Understood.

> We suggest using utest_printf() or greentea_send_kv() to send messages to the serial output, which will achieve the same results.

@jamesbeyond Should these be updated to utest_printf() ?

Reply:
>Should these be updated to utest_printf() ?

We'd suggest to utest_printf() over printf(), unless they are not interchangeable. In your case, I think the outcome would be the same.
Because it is less risk of disturbing the greentea communication strings, especially greentea is busy polling on serial. 
>Is this printf vs greentea kv / utest printf documented anywhere ? If not, can we add it. I've noticed this in some other tests also - often debug messages are left in tests.

Sorry, we don't have a formal document about that. We will add one in.

@0xc0170 @davidsaada @dannybenor